vapi/appliance/simulator/simulator.go

```diff
@@ -0,0 +1,177 @@
+			Enabled: false,
+			Timeout: 0,
+		},
+	}
+}
+
+// Register Appliance Management API paths with the vapi simulator's http.ServeMux
+func (h *Handler) Register(s *simulator.Service, r *simulator.Registry) {
+	s.HandleFunc(consolecli.Path, h.consoleCLI)
+	s.HandleFunc(dcui.Path, h.dcui)
+	s.HandleFunc(ssh.Path, h.ssh)
+	s.HandleFunc(shell.Path, h.shell)
+}
+
+func (h *Handler) decode(r *http.Request, w http.ResponseWriter, val interface{}) bool {
+	return Decode(r, w, val)
+}
+
+// Decode the request Body into val.
+// Returns true on success, otherwise false and sends the http.StatusBadRequest response.
+func Decode(r *http.Request, w http.ResponseWriter, val interface{}) bool {
+	defer r.Body.Close()
+	err := json.NewDecoder(r.Body).Decode(val)
+	if err != nil {
+		log.Printf("%s %s: %s", r.Method, r.RequestURI, err)
+		return false
+	}
+	return true
+}
+
+func (h *Handler) consoleCLI(writer http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodGet:
+		vapi.StatusOK(writer, h.consoleCLIAccess.Enabled)
+	case http.MethodPut:
+		var input consolecli.Access
+		if h.decode(request, writer, &input) {
+			go func() {
+				h.Lock()
+				h.consoleCLIAccess.Enabled = input.Enabled
+				h.Unlock()
+			}()
+			writer.WriteHeader(http.StatusNoContent)
+		} else {
+			writer.WriteHeader(http.StatusInternalServerError)
+		}
+	default:
+		http.NotFound(writer, request)
+	}
+}
+
+func (h *Handler) dcui(writer http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodGet:
+		vapi.StatusOK(writer, h.dcuiAccess.Enabled)
+	case http.MethodPut:
+		var input dcui.Access
+		if h.decode(request, writer, &input) {
+			go func() {
+				h.Lock()
+				h.dcuiAccess.Enabled = input.Enabled
+				h.Unlock()
+			}()
+			writer.WriteHeader(http.StatusNoContent)
+		} else {
+			writer.WriteHeader(http.StatusInternalServerError)
+		}
+	default:
+		http.NotFound(writer, request)
+	}
+}
+
+func (h *Handler) ssh(writer http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodGet:
+		vapi.StatusOK(writer, h.sshAccess.Enabled)
+	case http.MethodPut:
+		var input ssh.Access
+		if h.decode(request, writer, &input) {
+			go func() {
+				h.Lock()
+				h.sshAccess.Enabled = input.Enabled
+				h.Unlock()
+			}()
+			writer.WriteHeader(http.StatusNoContent)
+		} else {
+			writer.WriteHeader(http.StatusInternalServerError)
+		}
+	default:
+		http.NotFound(writer, request)
+	}
+}
+
+func (h *Handler) shell(writer http.ResponseWriter, request *http.Request) {
+	disableShell := func(delay int) {
+		go time.AfterFunc(time.Duration(delay)*time.Minute, func() {
+			h.Lock()
+			h.shellAccess.Timeout = 0
+			h.shellAccess.Enabled = false
+			h.Unlock()
+		})
+	}
+	switch request.Method {
+	case http.MethodGet:
+		vapi.StatusOK(writer, h.shellAccess)
+	case http.MethodPut:
+		var input shell.Access
+		if h.decode(request, writer, &input) {
+			h.shellAccess.Timeout = input.Timeout
```

Comment:
you don't use locks here, again we need to clarify if you need locks in each handler

Reply:
I have updated the code

```
func (h *Handler) shellStatus(writer http.ResponseWriter, request *http.Request) {
	setShell := func(timeout int, enabled bool) {
		h.Lock()
		h.shell.Timeout = timeout
		h.shell.Enabled = enabled
		h.Unlock()
	}
	switch request.Method {
	case http.MethodGet:
		vapi.StatusOK(writer, h.shell)
	case http.MethodPut:
		var input shell.Access
		if h.decode(request, writer, &input) {
			setShell(input.Timeout, input.Enabled)
			go time.AfterFunc(time.Duration(input.Timeout)*time.Minute, func() {
				setShell(0, false)
			})
			writer.WriteHeader(http.StatusNoContent)
		}
	default:
		http.NotFound(writer, request)
	}
}
```